feat: friends & chat redesign

lorux0 opened the pull request:
## What does this PR change?

All of the UI related to friends & chat feature has been changed with new visual design, flow implementations & architecture improvements.


## Friends

### Friends tab

All your friends should be displayed here, either they are online or offline.
Both lists are expandables/collapsables by clicking the header.
It also includes an empty state in case you have no friends online/offline or none added at all.
The online/offline friends should be sorted by the latest person you chatted with. 
When you hover the mouse over you one of your friends, a set of buttons should be displayed:
- jump in (in case in online)
- see profile
- whisper
- contextual menu which supports same options with the addition of block, report or unfriend

![Screenshot 2022-03-10 175638](https://user-images.githubusercontent.com/56365551/157754083-7b4c3bec-8935-412f-8505-e1373e602e14.jpg)
![Screenshot 2022-03-10 175656](https://user-images.githubusercontent.com/56365551/157755027-982f30e2-a0eb-46ba-ab53-32af4d22f897.jpg)

#### Search any friend

You can type any name or address of your friends and will be filtered providing the same options as the normal visualization.
Whenever you search, the friends should be sorted alphabetically in the list.

![Screenshot 2022-03-10 175815](https://user-images.githubusercontent.com/56365551/157755166-baac2c63-10df-4cc9-90ae-4bdf8e86758e.jpg)


### Requests tab

Here all the friend requests will be displayed, either ones you sent or received.
Both received and sent lists are expandable/collapsable by clicking the header.
To send a new request type the name or address of the person in the search bar and hit enter. A notification should be displayed telling you if it was sent correctly.

If you hover the mouse over a received request the following options should appear:
- add
- reject
- more options contextual menu

If you hover the mouse over a sent request the following options should appear:
- cancel
- more options contextual menu

![Screenshot 2022-03-10 175851](https://user-images.githubusercontent.com/56365551/157756332-8ac4cc5a-0957-4c92-a6b3-df89042d22b7.jpg)
![Screenshot 2022-03-10 180155](https://user-images.githubusercontent.com/56365551/157756360-5a2102cb-a2e7-4f59-b57d-855a697fba5f.jpg)

## Chat

### Channel list

In this window you can visualize all the private chats and public channels (#nearby only for now) that you have available.
Also you can search for any channel in the search bar.

Direct messages should contain any friend that you have chatted with at anytime. Friends that never got in any private conversation should not appear here. Every entry should be composed by:
- name
- last message received/sent
- when you hover with the mouse over the entry, a contextual menu option should appear: friendship status, profile, block, report, unfriend
- the unread messages count as a small notification
- click action should open a private conversation window

Public channels should only contain the #nearby chat for now. The entry should show:
- name
- the unread messages count as a small notification
- click action should open a public channel window

<img width="330" alt="Screen Shot 2022-05-05 at 12 22 44" src="https://user-images.githubusercontent.com/56365551/166957275-9816a680-0b5f-4586-a990-b3f438b3d7d7.png">

### Public channels

Display all the chats from nearby players with a max of 30. You should also see whispers in `From [player]: [message]` and `To [player]: [message]` formats depending if received or sent.

When you hover with the mouse over an entry, a toast should be displayed indicating the date. When the entry is a coordinate (for example `100,100`) you should see `100,100 INFO` format.

If you received any whisper, you should be able to reply with `/r ` filling automatically the name of the user.

You should be able to whisper a player typing `/w [playerName] message`.

When the window is opened, all the unread notifications from the channel should be cleared.

<img width="222" alt="Screen Shot 2022-05-05 at 14 57 56" src="https://user-images.githubusercontent.com/56365551/166984624-ea5a7bd9-2a0d-4215-8871-5571453267de.png">

You can right-click over a user's name to open a menu that should contain: profile, block, report options and, in case of a reigstered user, a friendship request option.

<img width="310" alt="Screen Shot 2022-05-05 at 15 00 33" src="https://user-images.githubusercontent.com/56365551/166984745-e6d04616-24d0-4827-b93e-f5bee8b81112.png">


### Direct messages

All private messages with a specific friend should be listed in this window with a max of 30.

When you hover with the mouse over an entry, a toast should be displayed indicating the date.

When the window is opened, all the notifications from that friend should be cleared.

In the header of the window should be displayed: the avatar's snapshot, name and, in case is online, a jump-in button. On the right, an options menu indicating: friendship status, profile, block, report, unfriend.


## Taskbar

The toggle on state of the buttons has been renewed.

The taskbar should display correctly the on/off state of the button depending on the window you have opened.

All unread chat notifications should be display over the chat button, including public channels and private channels.

All unread pending friend request notifications should be displayed over the friends button.
A new loading spinner has been added over the friends button while the service is loading.

<img width="202" alt="Screen Shot 2022-05-05 at 16 57 42" src="https://user-images.githubusercontent.com/56365551/167015662-4d49710f-9506-4c9f-8f26-a33d9634ea93.png">


## Kernel dependencies
- https://github.com/decentraland/kernel/pull/282
- https://github.com/decentraland/kernel/pull/296

## Release plan

The goal of the plan is to keep reporting social metrics at all times.

- [x] Merge and deploy https://github.com/decentraland/kernel/pull/282 into kernel's `main`
- [ ] Merge this PR into renderer's `master`
- [ ] Deploy to prod the latest version of the renderer
- [ ] Merge and deploy https://github.com/decentraland/kernel/pull/296

## Our Code Review Standards

https://github.com/decentraland/unity-renderer/blob/master/docs/code-review-standards.md

davidejensen commented on unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/FriendsHUD/Scripts/FriendsHUDView.cs:
```diff
@@ -31,14 +65,106 @@ public static FriendsHUDView Create(FriendsHUDController controller)
+    public void UpdateFriendshipStatus(string userId, FriendshipAction friendshipAction,
+        FriendEntryBase.Model friendEntryModel)
+    {
+        switch (friendshipAction)
```
Could it be worth having a default case which throws an error to simplify future cases in which new actions could be added?

lorux0 replied:
i will add an error log so it does not go unnoticed.. throwing an exception may break the flow if not handled properly

davidejensen commented on unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/FriendsHUD/Scripts/Tabs/CollapsableSortedFriendEntryList.cs:
```diff
@@ -0,0 +1,98 @@
+        filteredCount = 0;
+        var regex = new Regex(search, RegexOptions.IgnoreCase);
+
+        foreach (var entry in entries)
```
I've been told that it is better to use for instead of foreach due to better performances, maybe it's worth even here even if it's just the code for the filtering that isn't used too much

lorux0 replied:
done

davidejensen commented on unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/FriendsHUD/Scripts/Tabs/CollapsableSortedFriendEntryList.cs:
```diff
@@ -0,0 +1,98 @@
+
+        foreach (var entry in entries)
+        {
+            var isMatch = regex.IsMatch(entry.Key)
```
Do you think this could be improved? Imagining a search on 100+ friends it could become quite heavy applying a total of 300+ "IsMatch"

lorux0 replied:
good find! done

davidejensen commented:
In the .org the friends icon is bigger then in current production, is that normal?
.org and .zone:
![Screenshot 2022-03-21 at 11 18 23](https://user-images.githubusercontent.com/4254522/159243512-7da61ba5-96e1-4c7b-92b9-61307e885f71.png)

current prod:
![Screenshot 2022-03-21 at 11 27 40](https://user-images.githubusercontent.com/4254522/159243468-c11f8d70-6975-4e59-aaa3-bd4da359e248.png)

davidejensen commented on unity-renderer/Assets/UIComponents/Scripts/Components/SearchBar/SearchBarComponentModel.cs:
```diff
@@ -0,0 +1,8 @@
+[Serializable]
+public class SearchBarComponentModel : BaseComponentModel
+{
+    public float idleTimeToTriggerSearch = 1f;
```
To me 1 second felt a bit too long

lorux0 replied:
its the default value for all the search bars in the ui

davidejensen commented:
Left a couple of comments but approved due to suggestions being minor changes, awesome update to the friends list!

sandrade-dcl commented on unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/FriendsHUD/Scripts/FriendsHUDComponentView.cs:
```diff
@@ -0,0 +1,230 @@
+    {
+        base.Awake();
+        
+        friendsTabFocusButton.onClick.AddListener(() => FocusTab(0));
+        friendRequestsTabFocusButton.onClick.AddListener(() => FocusTab(1));
```
Could we create constants for these kind of indexes?

```suggestion
        friendsTabFocusButton.onClick.AddListener(() => FocusTab(FRIENDS_TAB_INDEX));
        friendRequestsTabFocusButton.onClick.AddListener(() => FocusTab(FRIEND_REQUESTS_INDEX));
```

sandrade-dcl commented on unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/FriendsHUD/Scripts/FriendsHUDComponentView.cs:
```diff
@@ -0,0 +1,230 @@
+
+    public void DisplayFriendUserNotFound() => friendRequestsTab.ShowUserNotFoundNotification();
+
+    public bool IsFriendListFocused() => friendsTab.gameObject.activeInHierarchy;
+
+    public bool IsFriendListCreationReady() => friendsTab.DidDeferredCreationCompleted;
```
Only as suggestion: As the returned data is a simple boolean, wouldn't it be better to declare these functions (`IsFriendListFocused()` and `IsFriendListCreationReady`) as properties? It is the same from the technical point of view, but maybe the code is a bit cleaner.

sandrade-dcl commented on unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/FriendsHUD/Scripts/FriendsHUDComponentView.cs:
```diff
@@ -0,0 +1,230 @@
+
+    public void Destroy() => Destroy(gameObject);
+
+    public void Show()
+    {
+        model.visible = true;
+        gameObject.SetActive(true);
+    }
+
+    public void Hide()
+    {
+        model.visible = false;
+        gameObject.SetActive(false);
+    }
```
The `BaseComponentView` class already have a `Show` and `Hide` function implemented. You should override them if you want to introduce a new implementation.

sandrade-dcl commented on unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/FriendsHUD/Scripts/FriendsHUDComponentView.cs:
```diff
@@ -0,0 +1,230 @@
+        friendsTab.Filter(search);
+    }
+
+    [Serializable]
+    private struct Model
+    {
+        public int focusedTabIndex;
+        public bool isLoadingSpinnerActive;
+        public bool visible;
+    }
```
Suggestion: Can we move the model struct to the beginning of the class?

sandrade-dcl commented on unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/FriendsHUD/Scripts/FriendsHUDController.cs:
```diff
@@ -26,15 +34,14 @@ public void Initialize(IFriendsController friendsController, UserProfile ownUser
-
-        view.friendsList.OnDeleteConfirmation += Entry_OnDelete;
+        
+        view.OnFriendRequestApproved += Entry_OnRequestAccepted;
+        view.OnCancelConfirmation += Entry_OnRequestCancelled;
+        view.OnRejectConfirmation += Entry_OnRequestRejected;
+        view.OnFriendRequestSent += Entry_OnRequestSent;
+        view.OnWhisper += Entry_OnWhisper;
+        view.OnDeleteConfirmation += Entry_OnDelete;
```
All these subscriptions are not being unsubscripted in the `Dispose`.

sandrade-dcl commented on unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/FriendsHUD/Scripts/FriendsHUDController.cs:
```diff
@@ -55,8 +62,12 @@ public void Initialize(IFriendsController friendsController, UserProfile ownUser
             }
         }
+        
+        chatController.OnAddMessage += HandleChatMessageAdded;
```
This subscription is not being unsubscripted in the Dispose.

sandrade-dcl commented on unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/FriendsHUD/Scripts/Tabs/CollapsableSortedFriendEntryList.cs:
```diff
@@ -0,0 +1,98 @@
+using DCL.Helpers;
+using UnityEngine;
+
+public class CollapsableSortedFriendEntryList : MonoBehaviour
```
As a collpasable list is something that is surely going to be used in future UI implementations, it could have been a good opportunity for creating a new UI Component (for example called `CollapsableContainer`) in our common library to be used here and wherever we want.

sandrade-dcl commented on unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/FriendsHUD/Scripts/Tabs/FriendRequestsTabComponentView.cs:
```diff
@@ -0,0 +1,308 @@
+    {
+        acceptedFriendNotification.model.timer = NOTIFICATIONS_DURATION;
+        acceptedFriendNotification.model.groupID = NOTIFICATIONS_ID;
+        acceptedFriendNotification.model.message = $"You and {requestEntry.model.userName} are now friends!";
```
Lets declare a constant for this kind of messages!

sandrade-dcl commented on unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/FriendsHUD/Scripts/Tabs/FriendRequestsTabView.cs:
```diff
@@ -105,6 +107,17 @@ public bool UpdateEntry(string userId, FriendEntry.Model model, bool? isReceived
+    
+    public void ShowRequestSuccessfullySentNotification()
+    {
+        requestSentNotification.model.message = $"Your request to {lastRequestSentUserName} successfully sent!";
```
Lets declare a constant for this kind of messages!

sandrade-dcl commented on unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/FriendsHUD/Scripts/Tabs/FriendsTabComponentView.cs:
```diff
@@ -0,0 +1,359 @@
+        offlineFriendsList.list.Expand();
+    }
+
+    public void Show()
+    {
+        gameObject.SetActive(true);
+        enabledHeader.SetActive(true);
+        disabledHeader.SetActive(false);
+    }
+
+    public void Hide()
+    {
+        gameObject.SetActive(false);
+        enabledHeader.SetActive(false);
+        disabledHeader.SetActive(true);
+        contextMenuPanel.Hide();
+    }
```
You should override the Show/Hide functions already implemented in the `BaseComponentView` class.

sandrade-dcl commented on unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/FriendsHUD/Scripts/Tabs/FriendsTabComponentView.cs:
```diff
@@ -0,0 +1,359 @@
+
+    private void UpdateCounterLabel()
+    {
+        onlineFriendsList.countText.SetText("ONLINE ({0})", onlineFriendsList.list.Count());
+        offlineFriendsList.countText.SetText("OFFLINE ({0})", offlineFriendsList.list.Count());
+        searchResultsFriendList.countText.SetText("Results ({0})", searchResultsFriendList.list.Count());
```
Lets declare constants for this kind of messages!

sandrade-dcl reviewed (approved):
Good work! I left several requests and suggestions but I think all of them are related to little details and shouldn't block my approve 👍

sandrade-dcl commented on unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/ChatWidgetHUD/ChatHUDView.cs:
```diff
@@ -1,128 +1,149 @@
+    {
+        add
+        {
+            if (value == null) return;
```
There could be any case where the subscription is null?

sandrade-dcl replied:
Of course, if it is not longer being used, lets delete it 👍

sandrade-dcl commented on unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/FriendsHUD/Scripts/Tabs/CollapsableSortedFriendEntryList.cs:
```diff
@@ -1,98 +1,13 @@
-        ((RectTransform) container).ForceUpdateLayout();
+        Filter(entry => regex.IsMatch(entry.model.userId)
+            || regex.IsMatch(entry.model.userName)
+            /*|| regex.IsMatch(entry.model.realm)*/);
```
Can we remove this commented line?

sandrade-dcl commented on unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/PrivateChatWindow/PrivateChatWindowComponentView.cs:
```diff
@@ -0,0 +1,110 @@
+        jumpInButtonContainer.SetActive(model.isUserOnline);
+    }
+
+    public void Setup(UserProfile profile, bool isOnline, bool isBlocked)
+    {
+        model = new Model
+        {
+            userId = profile.userId,
+            faceSnapshotUrl = profile.face256SnapshotURL,
+            userName = profile.userName,
+            isUserOnline = isOnline,
+            isUserBlocked = isBlocked
+        };
+        RefreshControl();
+    }
```
We already prepared, in our UI Components architecture, an interface called `IComponentModelConfig` to be used when we need to create a function for setting the model of our view as it is your case. Couldn't you use it for this case instead of declaring a `Setup()` method?

sandrade-dcl commented on unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/SocialBarHUD/LastReadMessagesService.cs:
```diff
@@ -0,0 +1,153 @@
+
+    public void Initialize()
+    {
+        if (chatController != null)
+            chatController.OnAddMessage += HandleMessageReceived;
+        LoadLastReadTimestamps();
```
```suggestion
        if (chatController != null)
            chatController.OnAddMessage += HandleMessageReceived;
            
        LoadLastReadTimestamps();
```

sandrade-dcl commented on unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/SocialBarHUD/LastReadMessagesService.cs:
```diff
@@ -0,0 +1,153 @@
+                    if (messageChannelId == ownUserId) return false;
+                    return messageChannelId == channelId
+                           && message.timestamp >= timestamp;
+                });;
```
```suggestion
                });
```

sandrade-dcl commented on unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/TaskbarHUD/TaskbarHUDController.cs:
```diff
@@ -112,177 +90,255 @@ public struct Configuration
             return;
 
-        OnFriendsToggleInputPress();
+        // ??????
```
What that comment means? If you think we should remark something here, could we change it by something more descriptive?

sandrade-dcl commented on unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/WorldChatWindowHUD/PrivateChatEntry.cs:
```diff
@@ -0,0 +1,115 @@
+        userContextMenu.OnBlock += HandleUserBlocked;
+    }
+
+    public void Set(PrivateChatEntryModel model)
+    {
+        this.model = model;
+        RefreshControl();
+    }
```
You can implement the interface `IComponentModelConfig` and replace this function by the `Configure(BaseComponentModel newModel)` one provided.

lorux0 replied:
done!

sandrade-dcl commented on unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/WorldChatWindowHUD/PublicChannelEntry.cs:
```diff
@@ -0,0 +1,56 @@
+        this.lastReadMessagesService = lastReadMessagesService;
+    }
+
+    public void Set(PublicChannelEntryModel model)
+    {
+        this.model = model;
+        RefreshControl();
+    }
```
You can implement the interface IComponentModelConfig and replace this function by the Configure(BaseComponentModel newModel) one provided.

lorux0 replied:
i think `IComponentModelConfig` should be generic `IComponentModelConfig<T>` since we are violating a LiskovSubstitution principle here

sandrade-dcl replied:
It would be ideal, Nico! I like the idea. Anyway, as I think this change would make us to modify a lot of other parts of code not related with your current implementation in this branch, in order to not increase the amount of changes here, I'm going to create a card task for that in order to apply this refactor in the future 👌

sandrade-dcl replied:
Done! Here is the task to be planned in future sprints: https://app.zenhub.com/workspaces/explorer-6047bad476c3c0001942ee91/issues/decentraland/unity-renderer/2291

sandrade-dcl commented on unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/WorldChatWindowHUD/PublicChatChannelComponentView.cs:
```diff
@@ -0,0 +1,58 @@
+
+    public void Show() => gameObject.SetActive(true);
+    
+    public void Setup(string channelId, string name, string description)
+    {
+        model.name = $"#{name}";
+        model.description = description;
+        RefreshControl();
+    }
```
You can implement the interface IComponentModelConfig and replace this function by the Configure(BaseComponentModel newModel) one provided.

lorux0 replied:
done

sandrade-dcl commented on unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/WorldChatWindowHUD/WorldChatWindowComponentView.cs:
```diff
@@ -0,0 +1,158 @@
+
+    public void Hide() => gameObject.SetActive(false);
+
+    public void SetPrivateChat(PrivateChatModel model)
+    {
+        var user = model.user;
+        directChatList.Set(user.userId, new PrivateChatEntry.PrivateChatEntryModel(
+            user.userId,
+            user.userName,
+            model.recentMessage.body,
+            user.face256SnapshotURL,
+            model.isBlocked,
+            model.isOnline,
+            model.recentMessage.timestamp));
+        directChatList.Sort();
+        UpdateHeaders();
+    }
```
You can implement the interface IComponentModelConfig and replace this function by the Configure(BaseComponentModel newModel) one provided.

lorux0 replied:
done!

sandrade-dcl commented on unity-renderer/Assets/UIComponents/Scripts/Components/CollapsableSortedList/CollapsableSortedListComponentView.cs:
```diff
@@ -0,0 +1,139 @@
+
+        public virtual int Count() => entries.Count - filteredCount;
+
+        public virtual void Show()
+        {
+            gameObject.SetActive(true);
+            model.isVisible = true;
+        }
+
+        public virtual void Hide()
+        {
+            gameObject.SetActive(false);
+            model.isVisible = false;
+        }
```
These `Show()` / `Hide()` functions... shouldn't be overriden from the `BaseComponentView` class?

lorux0 replied:
done!

sandrade-dcl reviewed (changes requested):
Very good work Nico! I left you some requests that I would like you to take a look before you can merge. Thanks!

sandrade-dcl commented:
I add some weird behaviours I have seen during the testing of the branch:

- **The ON/OFF status of the chat button in the taskbar is not being updated correctly**
1. We open the chat from the taskbar:
![image](https://user-images.githubusercontent.com/64659061/167312851-5948b884-18ab-4bd1-866c-6b3d46e3d258.png)
2. We click on the world to close the chat.
![image](https://user-images.githubusercontent.com/64659061/167312930-6555fafb-4b22-4fc2-a447-c5d3ba986b4b.png)
3. We click on the second button of the mouse to unblock it. The chat is re-opening but notice the status of the taskbar icon is not the correct.
![image](https://user-images.githubusercontent.com/64659061/167313064-6757e5f6-b14e-4d74-a00e-0bca75f8b9ea.png)

Anyway... I doin't know if the expected behaviour is to open the chat window each time we unblock the mouse. Is that correct? For me it is a bit weird, even annoying.

sandrade-dcl commented:
Is it normal that, if we click on on the row of a friend, nothing happens? To open the private chat we have to specifically click on the message icon in the right side of the row. If it is designed in this way, ok, although I would make it to open the private chat in both cases: when you click in the row and when you click on the icon.

![image](https://user-images.githubusercontent.com/64659061/167314028-4289a86c-a40f-4017-b91c-59c8d124d626.png)

sandrade-dcl commented:
It seems the margins in the welcome header of the general chat are a bit short.

![image](https://user-images.githubusercontent.com/64659061/167314149-1690bf4d-d88f-4ff2-bc4a-933311200fec.png)

RominaMarchetti commented:
@lorux0 I confirmed in Figma that you solved a few of the last comments and added a new ones (two to be honest and nothing critical).

In terms of usability:
- Be careful with what is happening in the search bar in Friends. It's is not being updated when you switch between sections after writing a name.
- About Santi's comment on the Friends Entry: if you click the container by default it should open the profile card, not the conversation.
- The key commands are broken. `enter` should open and close the #nearby conversation (ex general) if there is not any input. `L` should open/close Friends. Both are working to open the windows but not for closing them.

Overall, I think it's looking good to merge it into master a soon as possible. Amazing work, @lorux0 😍

lorux0 commented:
> > In the .org the friends icon is bigger then in current production, is that normal? .org and .zone: ![Screenshot 2022-03-21 at 11 18 23](https://user-images.githubusercontent.com/4254522/159243512-7da61ba5-96e1-4c7b-92b9-61307e885f71.png)
> > current prod: ![Screenshot 2022-03-21 at 11 27 40](https://user-images.githubusercontent.com/4254522/159243468-c11f8d70-6975-4e59-aaa3-bd4da359e248.png)
> 
> it will be fixed when i rebase the latest changes from master, since the taskbar has been modified by another pr

fixed it latest changes

lorux0 commented on unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/ChatWidgetHUD/ChatHUDView.cs:
```diff
@@ -166,27 +184,97 @@ public virtual void AddEntry(ChatEntry.Model chatEntryModel, bool setScrollPosit
+        entries.Remove(entries[0]);
+    }
+
+    public void Hide()
```
this is legacy code, this class is deprecated. I didnt delete it yet due merging conflicts, but i was thinking doing so, what do you think?

sandrade-dcl replied:
Yep!

lorux0 commented on unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/ChatWidgetHUD/ChatHUDView.cs:
```diff
@@ -166,27 +184,97 @@ public virtual void AddEntry(ChatEntry.Model chatEntryModel, bool setScrollPosit
+
+    public virtual void ClearAllEntries()
+    {
+        foreach (var entry in entries)
+            Destroy(entry.gameObject);
+        entries.Clear();
```
sounds good to me!

lorux0 commented on unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/ChatWidgetHUD/ChatHUDView.cs:
```diff
@@ -166,27 +184,97 @@ public virtual void AddEntry(ChatEntry.Model chatEntryModel, bool setScrollPosit
+
+    private bool IsEntryVisible(ChatEntry entry)
+    {
+        int visibleCorners =
+            (entry.transform as RectTransform).CountCornersVisibleFrom(scrollRect.viewport.transform as RectTransform);
+        return visibleCorners > 0;
```
this is legacy code, this class is deprecated. I didnt delete it yet due merging conflicts, but i was thinking doing so, what do you think?

sandrade-dcl replied:
Yep!

lorux0 commented on unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/ChatWidgetHUD/DefaultChatEntry.cs:
```diff
@@ -0,0 +1,323 @@
+
+        if (chatEntryModel.messageType == ChatMessage.Type.PRIVATE)
+        {
+            if (chatEntryModel.subType == ChatEntryModel.SubType.RECEIVED)
+            {
+                userString = $"<b><color=#5EBD3D>From {chatEntryModel.senderName}:</color></b>";
+            }
+            else if (chatEntryModel.subType == ChatEntryModel.SubType.SENT)
+            {
+                userString = $"<b>To {chatEntryModel.recipientName}:</b>";
+            }
```
i dont really find an improvement here, since the format is no repeated in many places.. this is how the code would look like, which i think it is less understandable:
```
private const string RECEIVED_WHISPER_FORMAT = "<b><color=#5EBD3D>From {0}:</color></b>";
....
userString = string.Format(RECEIVED_WHISPER_FORMAT, chatEntryModel.senderName);
```

lorux0 commented on unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/ChatWidgetHUD/DefaultChatEntry.cs:
```diff
@@ -0,0 +1,323 @@
+                PreloadSceneMetadata(CoordinateUtils.ParseCoordinatesString(textCoordinates[i]));
+
+                body.text = body.text.Replace(textCoordinates[i],
+                    $"</noparse><link={textCoordinates[i]}><color=#4886E3><u>{textCoordinates[i]}</u></color></link><noparse>");
```
what do you think? https://github.com/decentraland/unity-renderer/pull/1994#discussion_r868405647

lorux0 commented:
> I add some weird behaviours I have seen during the testing of the branch:
> 
> * **The ON/OFF status of the chat button in the taskbar is not being updated correctly**
> 
> 1. We open the chat from the taskbar:
>    ![image](https://user-images.githubusercontent.com/64659061/167312851-5948b884-18ab-4bd1-866c-6b3d46e3d258.png)
> 2. We click on the world to close the chat.
>    ![image](https://user-images.githubusercontent.com/64659061/167312930-6555fafb-4b22-4fc2-a447-c5d3ba986b4b.png)
> 3. We click on the second button of the mouse to unblock it. The chat is re-opening but notice the status of the taskbar icon is not the correct.
>    ![image](https://user-images.githubusercontent.com/64659061/167313064-6757e5f6-b14e-4d74-a00e-0bca75f8b9ea.png)
> 
> Anyway... I doin't know if the expected behaviour is to open the chat window each time we unblock the mouse. Is that correct? For me it is a bit weird, even annoying.

thanks, fixed!

lorux0 commented:
> It seems the margins in the welcome header of the general chat are a bit short.
> 
> ![image](https://user-images.githubusercontent.com/64659061/167314149-1690bf4d-d88f-4ff2-bc4a-933311200fec.png)

fixed!

lorux0 commented:
> Is it normal that, if we click on on the row of a friend, nothing happens? To open the private chat we have to specifically click on the message icon in the right side of the row. If it is designed in this way, ok, although I would make it to open the private chat in both cases: when you click in the row and when you click on the icon.
> 
> ![image](https://user-images.githubusercontent.com/64659061/167314028-4289a86c-a40f-4017-b91c-59c8d124d626.png)

i would need a confirmation from @RominaMarchetti here

lorux0 commented:
> 

fixed, it opens the profile card now!

lorux0 commented:
> Is it normal that, if we click on on the row of a friend, nothing happens? To open the private chat we have to specifically click on the message icon in the right side of the row. If it is designed in this way, ok, although I would make it to open the private chat in both cases: when you click in the row and when you click on the icon.
> 
> ![image](https://user-images.githubusercontent.com/64659061/167314028-4289a86c-a40f-4017-b91c-59c8d124d626.png)

It opens the profile card now!

lorux0 commented:
> @lorux0 I confirmed in Figma that you solved a few of the last comments and added a new ones (two to be honest and nothing critical).
> 
> In terms of usability:
> 
> * Be careful with what is happening in the search bar in Friends. It's is not being updated when you switch between sections after writing a name.
> * About Santi's comment on the Friends Entry: if you click the container by default it should open the profile card, not the conversation.
> * The key commands are broken. `enter` should open and close the #nearby conversation (ex general) if there is not any input. `L` should open/close Friends. Both are working to open the windows but not for closing them.
> 
> Overall, I think it's looking good to merge it into master a soon as possible. Amazing work, @lorux0 😍

fixed! not really sure about the behaviour when you have an open chat, lock/unlock the cursor and then press enter

pravusjif reviewed (changes requested):
Everything seems to work correctly, I only saw a problem with some "white blocks" rendering order regarding (probably) friends passport picture on top of the chats:

https://user-images.githubusercontent.com/1031741/168871865-ad4b8536-f122-412a-8f39-4248caa2d11d.mp4
